### DISCUSSION TITLE
Name of selected MooseEnum item

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#18398] [Q&A General] [answered]
I would like to use the name (string) of an item selected in a MooseEnum to be a suffix as part of another name.  Let me try to explain.  Given the declaration:

```
   const field::FieldEnum _field;
```

and the initialization in a Material constructor:

```
   _field(getParam<MooseEnum>("field").getEnum<field::FieldEnum>()),
```

I would then like to declare a property like:

```
   _alfa(declareADProperty<Real>("alfa_" + name_of(_field_)),
```

where name_of(_field) would be the name associated with the MooseEnum item _field.  _field is (of course) an integer and I have been unable to find how to get the associated name.  I am sure this is due to my minimal understanding of MooseEnum.

Thanks for the help,
  Joe Kelly

**Comment by GiudGiud** [comment]
Hi Joe

First do you need to do this:
`_field(getParam("field").getEnumfield::FieldEnum())`
Or could you get by with
`_field(getParam<MooseEnum>("field")),` 
?

This field enum is a structure you declared right? It inherits from MooseEnum right?

You want to use the name() attribute of the enumerated objects. You could retrieve the name using the MooseEnumBase::find(int ...) routine with _field. This might not be the most graceful way though.

Guillaume

**Reply by joe61vette** [reply]
Thanks Guillaume.  It appears that "find(int ...)" returns the id not the name.  So, I went ahead and added a method:

std::string getFieldName(int id)
{
  switch (id)
  {
    case 0:
      return "liquid";

    case 1:
      return "gas";

    case 2:
      return "drop";

    default:
      mooseError("Invalid id for FieldEnum, id = ", id);
  }

Just a bit surprised that there isn't a more generic way to do this in Moose as there seems to be multiple ways to do just about anything.

  Joe

**Reply by GiudGiud** [reply]
find with the id as argument should return an iterator which you should be able to use to call name(). 
```std::set<MooseEnumItem>::const_iterator
MooseEnumBase::find(int id) const
{
  return std::find_if(
      _items.begin(), _items.end(), [&id](MooseEnumItem const & item) { return item.id() == id; });
}

```
Another option is to call getNames() on the enum and access it at the right position.

**Reply by joe61vette** [reply]
I got the getNames() option to work.  Thanks!

**Comment by dschwen** [comment, marked as answer]
just do

```
_field_name(getParam<MooseEnum>("field"))
```

in your initializer list, where `_field_name` is a `const std::string` member.

**Reply by joe61vette** [reply]
Thanks Dan.  Could you please mark this as answered?

  Joe
